// @sajari/react-components/dist/Select/components/Button/styles.esm.js
import { extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { mapStyles } from '@sajari/react-sdk-utils';
import '../../../hooks/useBorderRadius.esm.js';
import '../../../hooks/useDropdownItemStyles.esm.js';
import '../../../hooks/useDropdownStyles.esm.js';
import '../../../hooks/useFirstRender.esm.js';
import '../../../hooks/useRingStyles.esm.js';
import '../../../hooks/useFocusRingStyles.esm.js';
import '../../../hooks/useFontSize.esm.js';
import useInputStyles from '../../../hooks/useInputStyles.esm.js';
import '../../../hooks/useJustifyContent.esm.js';
import '../../../hooks/useSpacing.esm.js';
import '../../../hooks/useVoiceInput.esm.js';
import 'use-resize-observer/polyfilled';

function useButtonStyles(params) {
  var _useInputStyles = useInputStyles(_extends({
    type: 'select'
  }, params)),
      inputStyles = _useInputStyles.styles,
      focusRingProps = _useInputStyles.focusProps,
      focusRingStyles = _useInputStyles.focusRingStyles;

  var styles = {
    container: [{
      "backgroundImage": "url(\"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 20 20' fill='none'%3e%3cpath d='M7 7l3-3 3 3m0 6l-3 3-3-3' stroke='%239fa6b2' stroke-width='1.5' stroke-linecap='round' stroke-linejoin='round'/%3e%3c/svg%3e\")",
      "appearance": "none",
      "colorAdjust": "exact",
      "backgroundRepeat": "no-repeat",
      "backgroundColor": "#ffffff",
      "borderColor": "#d2d6dc",
      "borderWidth": "1px",
      "borderRadius": "0.375rem",
      "paddingTop": "0.5rem",
      "paddingRight": "2.5rem",
      "paddingBottom": "0.5rem",
      "paddingLeft": "0.75rem",
      "fontSize": "1rem",
      "lineHeight": "1.5",
      "backgroundPosition": "right 0.5rem center",
      "backgroundSize": "1.5em 1.5em",
      "::-ms-expand": {
        "color": "#9fa6b2",
        "border": "none"
      },
      ":focus": {
        "outline": "none",
        "boxShadow": "0 0 0 3px rgba(164, 202, 254, 0.45)",
        "borderColor": "#a4cafe"
      },
      "@media not print": {
        "::-ms-expand": {
          "display": "none"
        }
      },
      "@media print and (-ms-high-contrast: active), print and (-ms-high-contrast: none)": {
        "paddingRight": "0.75rem"
      }
    }, inputStyles]
  };
  return {
    styles: mapStyles(styles),
    focusRingProps: focusRingProps,
    focusRingStyles: focusRingStyles
  };
}

export { useButtonStyles };
