// @sajari/react-hooks/dist/ContextProvider/controllers/Response.esm.js
var Response = /*#__PURE__*/function () {
  function Response(error, queryValues, response, values) {
    this.error = error;
    this.queryValues = queryValues;
    this.response = response;
    this.values = values;
  }

  var _proto = Response.prototype;

  _proto.isEmpty = function isEmpty() {
    return this.error === null && this.response === undefined && this.values === undefined && this.queryValues === undefined;
  };

  _proto.isError = function isError() {
    return this.error !== null;
  };

  _proto.getError = function getError() {
    return this.error;
  };

  _proto.getQueryValues = function getQueryValues() {
    return this.queryValues;
  };

  _proto.getResponse = function getResponse() {
    return this.response;
  };

  _proto.getValues = function getValues() {
    return this.values;
  };

  _proto.getResults = function getResults() {
    return this.response !== undefined ? this.response.get('results') : undefined;
  };

  _proto.getTotalResults = function getTotalResults() {
    return this.response !== undefined ? this.response.get('totalResults') : undefined;
  };

  _proto.getTime = function getTime() {
    return this.response !== undefined ? this.response.get('time') : undefined;
  };

  _proto.getAggregates = function getAggregates() {
    if (this.response === undefined) {
      return undefined;
    }

    var aggregates = this.response.get('aggregates');

    if (aggregates === undefined) {
      return undefined;
    }

    return aggregates;
  };

  _proto.getAggregateFilters = function getAggregateFilters() {
    if (this.response === undefined) {
      return undefined;
    }

    var aggregates = this.response.get('aggregateFilters');

    if (aggregates === undefined) {
      return undefined;
    }

    return aggregates;
  };

  return Response;
}();

export { Response };
