// @sajari/react-search-ui/dist/Filter/ColorFilter.esm.js
import { useFilter } from '@sajari/react-hooks';
import { isEmpty } from '@sajari/react-sdk-utils';
import { useMemo } from 'react';
import { jsx } from '@emotion/core/jsx-runtime';
import { useSearchUIContext } from '../ContextProvider/index.esm.js';
import { Swatch } from '@sajari/react-components';
import Box from './Box.esm.js';

var colorKeys = Swatch.colorKeys;

var ColorFilter = function ColorFilter(_ref) {
  var _customClassNames$fil, _customClassNames$fil2, _customClassNames$fil3, _customClassNames$fil4, _customClassNames$fil5, _customClassNames$fil6;

  var name = _ref.name,
      title = _ref.title;

  var _useFilter = useFilter(name),
      options = _useFilter.options,
      selected = _useFilter.selected,
      setSelected = _useFilter.setSelected,
      reset = _useFilter.reset;

  var _useSearchUIContext = useSearchUIContext(),
      customClassNames = _useSearchUIContext.customClassNames,
      _useSearchUIContext$d = _useSearchUIContext.disableDefaultStyles,
      disableDefaultStyles = _useSearchUIContext$d === void 0 ? false : _useSearchUIContext$d;

  var optionKeys = useMemo(function () {
    return options.map(function (o) {
      return o.label;
    });
  }, [JSON.stringify(options)]);
  var filtered = useMemo(function () {
    return colorKeys.filter(function (c) {
      return optionKeys.some(function (o) {
        return c.toLowerCase() === o.toLowerCase();
      });
    });
  }, [JSON.stringify(optionKeys)]);
  var children = useMemo(function () {
    return filtered.map(function (color) {
      var Component = Swatch.Color[color];
      return jsx(Component, {}, color);
    });
  }, [JSON.stringify(filtered)]);

  if (isEmpty(filtered) && isEmpty(selected)) {
    return null;
  }

  return jsx(Box, {
    title: title,
    showReset: selected.length > 0,
    onReset: reset,
    children: jsx(Swatch, {
      checkedColors: selected,
      onChange: setSelected,
      className: (_customClassNames$fil = customClassNames.filter) == null ? void 0 : (_customClassNames$fil2 = _customClassNames$fil.color) == null ? void 0 : _customClassNames$fil2.container,
      colorClassName: (_customClassNames$fil3 = customClassNames.filter) == null ? void 0 : (_customClassNames$fil4 = _customClassNames$fil3.color) == null ? void 0 : _customClassNames$fil4.item,
      colorCheckedClassName: (_customClassNames$fil5 = customClassNames.filter) == null ? void 0 : (_customClassNames$fil6 = _customClassNames$fil5.color) == null ? void 0 : _customClassNames$fil6.itemChecked,
      disableDefaultStyles: disableDefaultStyles,
      children: children
    })
  });
};

export default ColorFilter;
