// @sajari/react-search-ui/dist/Filter/RatingFilter.esm.js
import { useCallback } from 'react';
import { jsx } from '@emotion/core/jsx-runtime';
import { useSearchUIContext } from '../ContextProvider/index.esm.js';
import { Rating } from '@sajari/react-components';
import ListFilter from './ListFilter.esm.js';

var RatingFilter = function RatingFilter(_ref) {
  var _customClassNames$fil7;

  var name = _ref.name,
      title = _ref.title;

  var _useSearchUIContext = useSearchUIContext(),
      ratingMax = _useSearchUIContext.ratingMax,
      disableDefaultStyles = _useSearchUIContext.disableDefaultStyles,
      customClassNames = _useSearchUIContext.customClassNames;

  var renderRating = useCallback(function (v) {
    var _customClassNames$fil, _customClassNames$fil2, _customClassNames$fil3, _customClassNames$fil4, _customClassNames$fil5, _customClassNames$fil6;

    return jsx(Rating, {
      max: ratingMax,
      className: (_customClassNames$fil = customClassNames.filter) == null ? void 0 : (_customClassNames$fil2 = _customClassNames$fil.rating) == null ? void 0 : _customClassNames$fil2.container,
      activeRatingItemClassName: (_customClassNames$fil3 = customClassNames.filter) == null ? void 0 : (_customClassNames$fil4 = _customClassNames$fil3.rating) == null ? void 0 : _customClassNames$fil4.activeRatingItem,
      activeHalfRatingItemClassName: (_customClassNames$fil5 = customClassNames.filter) == null ? void 0 : (_customClassNames$fil6 = _customClassNames$fil5.rating) == null ? void 0 : _customClassNames$fil6.activeHalfRatingItem,
      value: Number(v),
      disableDefaultStyles: disableDefaultStyles
    });
  }, [ratingMax, JSON.stringify((_customClassNames$fil7 = customClassNames.filter) == null ? void 0 : _customClassNames$fil7.rating), disableDefaultStyles]);
  return jsx(ListFilter, {
    name: name,
    title: title,
    sort: "none",
    pinSelected: false,
    itemRender: renderRating
  });
};

export default RatingFilter;
