// @sajari/react-components/dist/Radio/styles.esm.js
import { extends as _extends } from '../_virtual/_rollupPluginBabelHelpers.js';
import { mapStyles } from '@sajari/react-sdk-utils';
import '../hooks/useBorderRadius.esm.js';
import '../hooks/useDropdownItemStyles.esm.js';
import '../hooks/useDropdownStyles.esm.js';
import '../hooks/useFirstRender.esm.js';
import '../hooks/useRingStyles.esm.js';
import '../hooks/useFocusRingStyles.esm.js';
import useFontSize from '../hooks/useFontSize.esm.js';
import useInputStyles from '../hooks/useInputStyles.esm.js';
import '../hooks/useJustifyContent.esm.js';
import '../hooks/useSpacing.esm.js';
import '../hooks/useVoiceInput.esm.js';
import 'use-resize-observer/polyfilled';

function useRadioStyles(props) {
  var _props$invalid = props.invalid,
      invalid = _props$invalid === void 0 ? false : _props$invalid,
      fontSize = props.fontSize;
  var sizeStyles = useFontSize({
    size: fontSize
  });

  var _useInputStyles = useInputStyles(_extends({
    type: 'radio'
  }, props)),
      inputStyles = _useInputStyles.styles,
      focusRingStyles = _useInputStyles.focusRingStyles,
      focusProps = _useInputStyles.focusProps;

  var styles = {
    container: [{
      "display": "flex",
      "alignItems": "center"
    }],
    label: [{
      "marginLeft": "0.5rem"
    }, sizeStyles],
    componentWrapper: [{
      "display": "inline-flex",
      "alignItems": "center"
    }],
    inputWrapper: [{
      "position": "relative",
      "display": "flex"
    }, focusRingStyles],
    input: [[{
      ":checked": {
        "backgroundImage": "url(\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3ccircle cx='8' cy='8' r='3'/%3e%3c/svg%3e\")",
        "borderColor": "transparent",
        "backgroundColor": "currentColor",
        "backgroundSize": "100% 100%",
        "backgroundPosition": "center",
        "backgroundRepeat": "no-repeat"
      },
      "appearance": "none",
      "colorAdjust": "exact",
      "display": "inline-block",
      "verticalAlign": "middle",
      "backgroundOrigin": "border-box",
      "userSelect": "none",
      "flexShrink": "0",
      "borderRadius": "100%",
      "height": "1rem",
      "width": "1rem",
      "color": "#3f83f8",
      "backgroundColor": "#ffffff",
      "borderColor": "#d2d6dc",
      "borderWidth": "1px",
      ":focus": {
        "outline": "none",
        "boxShadow": "0 0 0 3px rgba(164, 202, 254, 0.45)",
        "borderColor": "#a4cafe"
      },
      ":checked:focus": {
        "borderColor": "transparent"
      },
      "@media not print": {
        "::-ms-check": {
          "borderWidth": "1px",
          "color": "transparent",
          "background": "inherit",
          "borderColor": "inherit",
          "borderRadius": "inherit"
        }
      }
    }, inputStyles]]
  };

  if (invalid) {
    styles.label.push({
      "--text-opacity": "1",
      "color": "rgba(240, 82, 82, var(--text-opacity))"
    });
  }

  return {
    styles: mapStyles(styles),
    focusProps: focusProps
  };
}

export default useRadioStyles;
