// @sajari/react-components/dist/hooks/useDropdownStyles.esm.js
import { mapStyles } from '@sajari/react-sdk-utils';

function useDropdownStyles(_ref) {
  var _ref$shown = _ref.shown,
      shown = _ref$shown === void 0 ? true : _ref$shown;
  var styles = {
    container: [{
      "position": "absolute",
      "left": "0",
      "right": "0",
      "top": "auto",
      "zIndex": "20",
      "marginTop": "0.25rem",
      "--bg-opacity": "0.95",
      "backgroundColor": "rgba(255, 255, 255, var(--bg-opacity))",
      "borderRadius": "0.375rem",
      "boxShadow": "inset 0 0 0 1px rgba(0, 0, 0, 0.1), 0 1px 5px rgba(0, 0, 0, 0.1), inset 0 -1px 0 rgba(0, 0, 0, 0.1)",
      "backdropFilter": "blur(0.25rem)"
    }, !shown ? {
      "display": "none"
    } : {}],
    heading: [{
      "paddingLeft": "1rem",
      "paddingRight": "1rem",
      "paddingTop": "0.5rem",
      "fontSize": "0.75rem",
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))",
      "fontWeight": "400"
    }],
    items: [{
      "margin": "0",
      "paddingTop": "0.25rem",
      "paddingBottom": "0.25rem",
      "paddingLeft": "0.5rem",
      "paddingRight": "0.5rem",
      "listStyleType": "none"
    }],
    footer: [{
      "display": "flex",
      "alignItems": "center",
      "justifyContent": "space-between",
      "paddingLeft": "1rem",
      "paddingRight": "1rem",
      "paddingTop": "0.5rem",
      "paddingBottom": "0.5rem",
      "fontSize": "0.75rem",
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))",
      "borderWidth": "1px",
      "--border-opacity": "1",
      "borderColor": "rgba(229, 231, 235, var(--border-opacity))",
      "borderBottomLeftRadius": "inherit",
      "borderBottomRightRadius": "inherit",
      "--bg-opacity": "1",
      "backgroundColor": "rgba(249, 250, 251, var(--bg-opacity))"
    }],
    footerItem: [{
      "display": "flex",
      "alignItems": "center"
    }],
    footerItems: [{
      "display": "flex",
      "alignItems": "center",
      "> :not(template) ~ :not(template)": {
        "--space-x-reverse": 0,
        "marginRight": "calc(1rem * var(--space-x-reverse))",
        "marginLeft": "calc(1rem * calc(1 - var(--space-x-reverse)))"
      }
    }],
    footerIcon: [{
      "marginRight": "0.375rem",
      "flexShrink": "0"
    }],
    footerEscHint: [{
      "marginRight": "0.25rem",
      "fontWeight": "500"
    }],
    poweredBy: [{
      "marginLeft": "auto"
    }]
  };
  return mapStyles(styles);
}

export { useDropdownStyles };
