// @sajari/react-components/dist/Pagination/styles.esm.js
import { mapStyles } from '@sajari/react-sdk-utils';
import '../hooks/useBorderRadius.esm.js';
import '../hooks/useDropdownItemStyles.esm.js';
import '../hooks/useDropdownStyles.esm.js';
import '../hooks/useFirstRender.esm.js';
import '../hooks/useRingStyles.esm.js';
import '../hooks/useFocusRingStyles.esm.js';
import '../hooks/useFontSize.esm.js';
import '../hooks/useInputStyles.esm.js';
import useJustifyContent from '../hooks/useJustifyContent.esm.js';
import '../hooks/useSpacing.esm.js';
import '../hooks/useVoiceInput.esm.js';
import 'use-resize-observer/polyfilled';

function usePaginationStyles(props) {
  var _props$align = props.align,
      align = _props$align === void 0 ? 'center' : _props$align;
  var justifyContentStyles = useJustifyContent({
    align: align
  });
  var spacerStyles = {
    "paddingTop": "0.5rem",
    "paddingBottom": "0.5rem",
    "borderWidth": "1px",
    "--border-opacity": "1",
    "borderColor": "rgba(229, 231, 235, var(--border-opacity))",
    "borderStyle": "solid",
    "borderRadius": "0",
    "userSelect": "none"
  };
  var styles = {
    container: [{
      "display": "flex"
    }, justifyContentStyles],
    spacerEllipsis: [{
      "paddingLeft": "0.5rem",
      "paddingRight": "0.5rem",
      "--bg-opacity": "1",
      "backgroundColor": "rgba(249, 250, 251, var(--bg-opacity))"
    }, spacerStyles],
    compactStatus: [{
      "paddingLeft": "1rem",
      "paddingRight": "1rem",
      "--bg-opacity": "1",
      "backgroundColor": "rgba(255, 255, 255, var(--bg-opacity))"
    }, spacerStyles],
    compactStatusCount: [{
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))"
    }]
  };
  return mapStyles(styles);
}

export default usePaginationStyles;
