// @sajari/react-components/dist/Tabs/TabPanel/index.esm.js
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import { assignRef, __DEV__ } from '@sajari/react-sdk-utils';
import { forwardRef } from 'react';
import { jsx } from '@emotion/core/jsx-runtime';
import Box from '../../Box/index.esm.js';
import { useTabContext } from '../context.esm.js';

var TabPanel = /*#__PURE__*/forwardRef(function (props, _ref) {
  var children = props.children,
      selected = props.selected,
      selectedPanelRef = props.selectedPanelRef,
      id = props.id,
      stylesProp = props.styles,
      rest = _objectWithoutPropertiesLoose(props, ["children", "selected", "selectedPanelRef", "id", "styles"]);

  var _useTabContext = useTabContext(),
      alwaysRenderChildren = _useTabContext.alwaysRenderChildren;

  return jsx(Box, _extends({
    ref: function ref(node) {
      if (selected && selectedPanelRef) {
        assignRef(selectedPanelRef, node);
      }

      assignRef(_ref, node);
    },
    role: "tabpanel",
    tabIndex: -1,
    "aria-labelledby": "tab:" + id,
    hidden: !selected,
    id: "panel:" + id,
    css: stylesProp
  }, rest, {
    children: (alwaysRenderChildren || selected) && children
  }));
});

if (__DEV__) {
  TabPanel.displayName = 'TabPanel';
}

export default TabPanel;
