// @sajari/react-search-ui/dist/Pagination/index.esm.js
import { usePagination } from '@sajari/react-hooks';
import { __DEV__ } from '@sajari/react-sdk-utils';
import 'react';
import { useTranslation } from 'react-i18next';
import { jsx } from '@emotion/core/jsx-runtime';
import { useSearchUIContext } from '../ContextProvider/index.esm.js';
import { extends as _extends } from '../_virtual/_rollupPluginBabelHelpers.js';
import { Pagination as Pagination$1 } from '@sajari/react-components';

var Pagination = function Pagination(props) {
  var _customClassNames$pag, _customClassNames$pag2, _customClassNames$pag3, _customClassNames$pag4, _customClassNames$pag5, _customClassNames$pag6, _customClassNames$pag7;

  var _useTranslation = useTranslation('pagination'),
      t = _useTranslation.t;

  var _usePagination = usePagination('search'),
      page = _usePagination.page,
      setPage = _usePagination.setPage,
      pageCount = _usePagination.pageCount,
      resultsPerPage = _usePagination.resultsPerPage,
      totalResults = _usePagination.totalResults;

  var _useSearchUIContext = useSearchUIContext(),
      customClassNames = _useSearchUIContext.customClassNames,
      disableDefaultStyles = _useSearchUIContext.disableDefaultStyles,
      language = _useSearchUIContext.language;

  return jsx(Pagination$1, _extends({
    language: language,
    page: page,
    resultsPerPage: resultsPerPage,
    totalResults: totalResults,
    pageCount: pageCount,
    onChange: setPage,
    i18n: {
      label: t('label'),
      previous: t('previous'),
      next: t('next'),
      page: function page(params) {
        return t('page', params);
      },
      current: function current(params) {
        return t('current', params);
      }
    },
    className: customClassNames == null ? void 0 : (_customClassNames$pag = customClassNames.pagination) == null ? void 0 : _customClassNames$pag.container,
    buttonClassName: customClassNames == null ? void 0 : (_customClassNames$pag2 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag2.button,
    activeClassName: customClassNames == null ? void 0 : (_customClassNames$pag3 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag3.active,
    prevClassName: customClassNames == null ? void 0 : (_customClassNames$pag4 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag4.prev,
    nextClassName: customClassNames == null ? void 0 : (_customClassNames$pag5 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag5.next,
    spacerEllipsisClassName: customClassNames == null ? void 0 : (_customClassNames$pag6 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag6.spacerEllipsis,
    statusClassName: customClassNames == null ? void 0 : (_customClassNames$pag7 = customClassNames.pagination) == null ? void 0 : _customClassNames$pag7.status,
    disableDefaultStyles: disableDefaultStyles
  }, props));
};

if (__DEV__) {
  Pagination.displayName = 'Pagination';
}

export default Pagination;
