// @sajari/react-components/dist/Checkbox/styles.esm.js
import { extends as _extends } from '../_virtual/_rollupPluginBabelHelpers.js';
import { mapStyles, useTheme } from '@sajari/react-sdk-utils';
import '../hooks/useBorderRadius.esm.js';
import '../hooks/useDropdownItemStyles.esm.js';
import '../hooks/useDropdownStyles.esm.js';
import '../hooks/useFirstRender.esm.js';
import '../hooks/useRingStyles.esm.js';
import '../hooks/useFocusRingStyles.esm.js';
import useFontSize from '../hooks/useFontSize.esm.js';
import useInputStyles from '../hooks/useInputStyles.esm.js';
import '../hooks/useJustifyContent.esm.js';
import '../hooks/useSpacing.esm.js';
import '../hooks/useVoiceInput.esm.js';
import 'use-resize-observer/polyfilled';

function useCheckboxStyles(props) {
  var theme = useTheme();
  var _props$invalid = props.invalid,
      invalid = _props$invalid === void 0 ? false : _props$invalid,
      _props$indeterminate = props.indeterminate,
      indeterminate = _props$indeterminate === void 0 ? false : _props$indeterminate,
      fontSize = props.fontSize;
  var sizeStyles = useFontSize({
    size: fontSize
  });

  var _useInputStyles = useInputStyles(_extends({
    type: 'checkbox',
    indeterminate: indeterminate
  }, props)),
      inputStyles = _useInputStyles.styles,
      focusRingStyles = _useInputStyles.focusRingStyles,
      focusProps = _useInputStyles.focusProps;

  var styles = {
    container: [{
      "display": "flex",
      "alignItems": "center"
    }],
    componentWrapper: [{
      "position": "relative",
      "display": "inline-flex",
      "alignItems": "center"
    }],
    inputWrapper: [{
      "position": "relative",
      "display": "flex"
    }, focusRingStyles],
    label: [{
      "marginLeft": "0.5rem"
    }, sizeStyles],
    indeterminate: [{
      "position": "absolute",
      "top": "0",
      "right": "0",
      "bottom": "0",
      "left": "0",
      "display": "flex",
      "alignItems": "center",
      "justifyContent": "center",
      "pointerEvents": "none"
    }],
    indeterminateInner: [{
      "margin": "auto",
      "width": "50%",
      "height": "0.125rem",
      "borderRadius": "0.125rem"
    }, {
      backgroundColor: theme.color.primary.text
    }],
    input: [{
      ":checked": {
        "backgroundImage": "url(\"data:image/svg+xml,%3csvg viewBox='0 0 16 16' fill='white' xmlns='http://www.w3.org/2000/svg'%3e%3cpath d='M5.707 7.293a1 1 0 0 0-1.414 1.414l2 2a1 1 0 0 0 1.414 0l4-4a1 1 0 0 0-1.414-1.414L7 8.586 5.707 7.293z'/%3e%3c/svg%3e\")",
        "borderColor": "transparent",
        "backgroundColor": "currentColor",
        "backgroundSize": "100% 100%",
        "backgroundPosition": "center",
        "backgroundRepeat": "no-repeat"
      },
      "appearance": "none",
      "colorAdjust": "exact",
      "display": "inline-block",
      "verticalAlign": "middle",
      "backgroundOrigin": "border-box",
      "userSelect": "none",
      "flexShrink": "0",
      "height": "1rem",
      "width": "1rem",
      "color": "#3f83f8",
      "backgroundColor": "#ffffff",
      "borderColor": "#d2d6dc",
      "borderWidth": "1px",
      "borderRadius": "0.25rem",
      ":focus": {
        "outline": "none",
        "boxShadow": "0 0 0 3px rgba(164, 202, 254, 0.45)",
        "borderColor": "#a4cafe"
      },
      ":checked:focus": {
        "borderColor": "transparent"
      },
      "@media not print": {
        "::-ms-check": {
          "borderWidth": "1px",
          "color": "transparent",
          "background": "inherit",
          "borderColor": "inherit",
          "borderRadius": "inherit"
        }
      }
    }, inputStyles]
  };

  if (invalid) {
    styles.label.push({
      "--text-opacity": "1",
      "color": "rgba(240, 82, 82, var(--text-opacity))"
    });
  }

  return {
    styles: mapStyles(styles),
    focusProps: focusProps
  };
}

export { useCheckboxStyles };
