// @sajari/react-components/dist/Swatch/components/Color/styles.esm.js
import { mapStyles } from '@sajari/react-sdk-utils';
import useBorderRadius from '../../../hooks/useBorderRadius.esm.js';
import '../../../hooks/useDropdownItemStyles.esm.js';
import '../../../hooks/useDropdownStyles.esm.js';
import '../../../hooks/useFirstRender.esm.js';
import '../../../hooks/useRingStyles.esm.js';
import useFocusRingStyles from '../../../hooks/useFocusRingStyles.esm.js';
import '../../../hooks/useFontSize.esm.js';
import '../../../hooks/useInputStyles.esm.js';
import '../../../hooks/useJustifyContent.esm.js';
import '../../../hooks/useSpacing.esm.js';
import '../../../hooks/useVoiceInput.esm.js';
import 'use-resize-observer/polyfilled';

function useColorStyles(props) {
  var checked = props.checked,
      border = props.border,
      rounded = props.rounded;

  var _useFocusRingStyles = useFocusRingStyles({
    color: String(border)
  }),
      focusProps = _useFocusRingStyles.focusProps,
      focusRingStyles = _useFocusRingStyles.focusRingStyles;

  var borderRadiusStyles = useBorderRadius({
    rounded: rounded
  });
  var styles = {
    container: [focusRingStyles, {
      "position": "relative",
      "display": "flex",
      "alignItems": "center",
      "justifyContent": "center",
      "marginTop": "0.5rem",
      "marginLeft": "0.5rem",
      "borderWidth": "1px",
      "borderStyle": "solid",
      "cursor": "pointer",
      "width": "1.75rem",
      "height": "1.75rem"
    }, borderRadiusStyles],
    iconCheck: [checked ? {
      "opacity": "1",
      "fill": "currentColor"
    } : {
      "opacity": "0",
      "fill": "currentColor"
    }]
  };
  return {
    styles: mapStyles(styles),
    focusProps: focusProps
  };
}

export default useColorStyles;
