// @sajari/react-components/dist/Tabs/TabPanels/index.esm.js
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import { cleanChildren, __DEV__ } from '@sajari/react-sdk-utils';
import React__default, { cloneElement } from 'react';
import { jsx } from '@emotion/core/jsx-runtime';
import Box from '../../Box/index.esm.js';
import { useTabContext } from '../context.esm.js';

var TabPanels = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var children = props.children,
      stylesProp = props.styles,
      rest = _objectWithoutPropertiesLoose(props, ["children", "styles"]);

  var _useTabContext = useTabContext(),
      selectedIndex = _useTabContext.index,
      selectedPanelRef = _useTabContext.selectedPanelRef,
      id = _useTabContext.id,
      manual = _useTabContext.manual,
      manualIndex = _useTabContext.manualIndex;

  var validChildren = cleanChildren(children);
  var clones = validChildren.map(function (child, index) {
    return /*#__PURE__*/cloneElement(child, {
      selected: manual ? index === manualIndex : index === selectedIndex,
      selectedPanelRef: selectedPanelRef,
      id: id + "-" + index
    });
  });
  return jsx(Box, _extends({
    tabIndex: -1,
    ref: ref,
    css: stylesProp
  }, rest, {
    children: clones
  }));
});

if (__DEV__) {
  TabPanels.displayName = 'TabPanels';
}

export default TabPanels;
