// @sajari/react-search-ui/dist/Sorting/index.esm.js
import { useSorting } from '@sajari/react-hooks';
import { isArray } from '@sajari/react-sdk-utils';
import 'react';
import { useTranslation } from 'react-i18next';
import { jsx } from '@emotion/core/jsx-runtime';
import { useSearchUIContext } from '../ContextProvider/index.esm.js';
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose, extends as _extends } from '../_virtual/_rollupPluginBabelHelpers.js';
import { Select, Option } from '@sajari/react-components';
import { useId } from '@react-aria/utils';
import ViewOption from '../ViewOption/index.esm.js';

var defaultOptions = [{
  name: 'Most relevant',
  value: ''
}];

var Sorting = function Sorting(props) {
  var _customClassNames$sor, _customClassNames$sor2, _customClassNames$sor3;

  var _useTranslation = useTranslation('sorting'),
      t = _useTranslation.t;

  var _props$label = props.label,
      label = _props$label === void 0 ? t('label') : _props$label,
      _props$options = props.options,
      options = _props$options === void 0 ? defaultOptions : _props$options,
      size = props.size,
      rest = _objectWithoutPropertiesLoose(props, ["label", "options", "size", "styles"]);

  var _useSearchUIContext = useSearchUIContext(),
      _useSearchUIContext$d = _useSearchUIContext.disableDefaultStyles,
      disableDefaultStyles = _useSearchUIContext$d === void 0 ? false : _useSearchUIContext$d,
      customClassNames = _useSearchUIContext.customClassNames;

  var _useSorting = useSorting(),
      sorting = _useSorting.sorting,
      setSorting = _useSorting.setSorting;

  var id = "sorting-" + useId();
  return jsx(ViewOption, _extends({
    id: id,
    label: label,
    size: size,
    className: (_customClassNames$sor = customClassNames.sorting) == null ? void 0 : _customClassNames$sor.container,
    labelClassName: (_customClassNames$sor2 = customClassNames.sorting) == null ? void 0 : _customClassNames$sor2.label,
    renderAsLabel: true
  }, rest, {
    children: jsx(Select, {
      id: id,
      value: sorting,
      onChange: function onChange(value) {
        return setSorting(isArray(value) ? value[0] : value);
      },
      size: size,
      disableDefaultStyles: disableDefaultStyles,
      className: (_customClassNames$sor3 = customClassNames.sorting) == null ? void 0 : _customClassNames$sor3.select,
      children: options.map(function (s) {
        return jsx(Option, {
          value: s.value,
          children: s.name
        }, s.value);
      })
    })
  }));
};

export default Sorting;
