// @sajari/react-search-ui/dist/Input/index.esm.js
import { useSearchContext, useAutocomplete, useQuery } from '@sajari/react-hooks';
import { isArray, __DEV__ } from '@sajari/react-sdk-utils';
import React__default, { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { jsx } from '@emotion/core/jsx-runtime';
import { useSearchUIContext } from '../ContextProvider/index.esm.js';
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose, extends as _extends } from '../_virtual/_rollupPluginBabelHelpers.js';
import { Combobox } from '@sajari/react-components';
import useClickTracking from '../hooks/useClickTracking.esm.js';
import mapResultFields from '../utils/mapResultFields.esm.js';

var Input = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  var _customClassNames$inp, _customClassNames$inp2, _customClassNames$inp3, _customClassNames$inp4, _customClassNames$inp5, _customClassNames$inp6, _customClassNames$inp7, _customClassNames$inp8, _customClassNames$inp9, _customClassNames$inp10, _customClassNames$inp11, _customClassNames$inp12, _customClassNames$inp13, _customClassNames$inp14, _customClassNames$inp15;

  var _useTranslation = useTranslation('input'),
      t = _useTranslation.t;

  var _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? t('placeholder') : _props$placeholder,
      _props$mode = props.mode,
      mode = _props$mode === void 0 ? 'instant' : _props$mode,
      onSelect = props.onSelect,
      onChange = props.onChange,
      rest = _objectWithoutPropertiesLoose(props, ["placeholder", "mode", "onSelect", "onChange"]);

  var _useSearchContext = useSearchContext(),
      rawResults = _useSearchContext.results,
      search = _useSearchContext.search,
      searching = _useSearchContext.searching,
      fields = _useSearchContext.fields;

  var results = React__default.useMemo(function () {
    return mapResultFields(rawResults != null ? rawResults : [], fields);
  }, [rawResults]);

  var _useAutocomplete = useAutocomplete(),
      searchAutocomplete = _useAutocomplete.search,
      completion = _useAutocomplete.completion,
      suggestions = _useAutocomplete.suggestions;

  var _useSearchUIContext = useSearchUIContext(),
      customClassNames = _useSearchUIContext.customClassNames,
      _useSearchUIContext$d = _useSearchUIContext.disableDefaultStyles,
      disableDefaultStyles = _useSearchUIContext$d === void 0 ? false : _useSearchUIContext$d,
      tracking = _useSearchUIContext.tracking;

  var _useQuery = useQuery(),
      query = _useQuery.query;

  var items = [];

  if (mode === 'suggestions') {
    items = suggestions;
  } else if (mode === 'results') {
    items = results.splice(0, 5).map(function (result) {
      var values = result.values,
          token = result.token;
      var description = values.description,
          image = values.image,
          title = values.title;

      var _useClickTracking = useClickTracking({
        tracking: tracking,
        values: values,
        token: token
      }),
          href = _useClickTracking.href,
          onClick = _useClickTracking.onClick;

      return {
        title: title,
        url: href,
        onClick: onClick,
        description: description,
        image: isArray(image) ? image[0] : image
      };
    });
  }

  var onChangeMemoized = useCallback(function (value) {
    if (onChange) {
      onChange(value);
    }

    if (mode === 'suggestions' || mode === 'typeahead') {
      if (searchAutocomplete) {
        searchAutocomplete(value);
      }
    } else if (mode === 'instant' || mode === 'results') {
      search(value);
    }
  }, [onChange, mode, search]);
  var onKeyDownMemoized = useCallback(function (e) {
    if (e.key === 'Enter' && (mode === 'typeahead' || mode === 'suggestions' || mode === 'standard')) {
      search(e.target.value);
    }
  }, [mode, search]);
  var onSelectMemoized = useCallback(function (item) {
    if (onSelect) {
      onSelect(item);
    }

    if (mode === 'suggestions') {
      search(item);
    }
  }, [mode, search]);
  return jsx(Combobox, _extends({}, rest, {
    ref: ref,
    placeholder: placeholder,
    mode: mode,
    items: items,
    completion: mode === 'typeahead' ? completion : '',
    value: query,
    loading: searching,
    onChange: onChangeMemoized,
    onKeyDown: onKeyDownMemoized,
    onSelect: onSelectMemoized,
    className: (_customClassNames$inp = customClassNames.input) == null ? void 0 : _customClassNames$inp.container,
    dropdownClassName: (_customClassNames$inp2 = customClassNames.input) == null ? void 0 : _customClassNames$inp2.dropdown,
    dropdownFooterClassName: (_customClassNames$inp3 = customClassNames.input) == null ? void 0 : _customClassNames$inp3.dropdownFooter,
    dropdownHighlightItemClassName: (_customClassNames$inp4 = customClassNames.input) == null ? void 0 : _customClassNames$inp4.dropdownHighlightItem,
    dropdownSelectedItemClassName: (_customClassNames$inp5 = customClassNames.input) == null ? void 0 : _customClassNames$inp5.dropdownSelectedItem,
    dropdownItemClassName: (_customClassNames$inp6 = customClassNames.input) == null ? void 0 : _customClassNames$inp6.dropdownItem,
    dropdownListClassName: (_customClassNames$inp7 = customClassNames.input) == null ? void 0 : _customClassNames$inp7.dropdownList,
    inputClassName: (_customClassNames$inp8 = customClassNames.input) == null ? void 0 : _customClassNames$inp8.input,
    inputContainerClassName: (_customClassNames$inp9 = customClassNames.input) == null ? void 0 : _customClassNames$inp9.inputContainer,
    resultClassName: (_customClassNames$inp10 = customClassNames.input) == null ? void 0 : _customClassNames$inp10.result,
    resultImageContainerClassName: (_customClassNames$inp11 = customClassNames.input) == null ? void 0 : _customClassNames$inp11.resultImageContainer,
    resultTextContainerClassName: (_customClassNames$inp12 = customClassNames.input) == null ? void 0 : _customClassNames$inp12.resultTextContainer,
    selectedResultClassName: (_customClassNames$inp13 = customClassNames.input) == null ? void 0 : _customClassNames$inp13.selectedResult,
    typeaheadClassName: (_customClassNames$inp14 = customClassNames.input) == null ? void 0 : _customClassNames$inp14.typeahead,
    voiceInputClassName: (_customClassNames$inp15 = customClassNames.input) == null ? void 0 : _customClassNames$inp15.voiceInput,
    disableDefaultStyles: disableDefaultStyles
  }));
});

if (__DEV__) {
  Input.displayName = 'Input';
}

export default Input;
