// @sajari/react-components/dist/RangeInput/styles.esm.js
import { mapStyles } from '@sajari/react-sdk-utils';

function useRangeInputStyles(props) {
  var isSingleHandle = props.isSingleHandle;
  var styles = {
    container: [{
      "display": "flex",
      "flexDirection": "column",
      "paddingLeft": "0.5rem",
      "paddingRight": "0.5rem"
    }],
    wrapper: [{
      "width": "100%",
      "marginTop": "2rem",
      "marginBottom": "0.75rem"
    }],
    ticks: [{
      "position": "relative"
    }],
    tickItem: [{
      "marginBottom": "0.625rem",
      "fontSize": "0.75rem",
      "textAlign": "center",
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))",
      ":after": {
        "content": "\"\"",
        "width": "0",
        "borderWidth": "0",
        "borderRightWidth": "1px",
        "--border-opacity": "1",
        "borderColor": "rgba(210, 214, 220, var(--border-opacity))",
        "borderStyle": "solid",
        "height": "0.375rem",
        "position": "absolute"
      }
    }, {
      bottom: '100%'
    }, '&::after { left: 50%; top: 100% }'],
    input: [{
      "display": "flex",
      "flexDirection": "column",
      "alignItems": "center",
      "@media (min-width: 640px)": {
        "flexDirection": "row"
      }
    }, isSingleHandle ? {
      "justifyContent": "flex-end"
    } : {
      "justifyContent": "space-between"
    }]
  };
  return mapStyles(styles);
}

export default useRangeInputStyles;
