// @sajari/react-search-ui/dist/Results/components/Message/styles.esm.js
import { mapStyles } from '@sajari/react-sdk-utils';

function useMessageStyles() {
  var styles = {
    container: [{
      "width": "100%",
      "paddingLeft": "1.5rem",
      "paddingRight": "1.5rem",
      "paddingTop": "5rem",
      "paddingBottom": "5rem",
      "textAlign": "center",
      "@media (min-width: 768px)": {
        "paddingTop": "10rem",
        "paddingBottom": "10rem",
        "paddingLeft": "3rem",
        "paddingRight": "3rem"
      }
    }],
    loadingWrapper: [{
      "--text-opacity": "1",
      "color": "rgba(107, 114, 128, var(--text-opacity))"
    }],
    loadingSpinner: [{
      "display": "inline-block",
      "width": "1.5rem",
      "height": "1.5rem"
    }],
    loadingText: [{
      "marginTop": "0.75rem"
    }],
    errorHeading: [{
      "--text-opacity": "1",
      "color": "rgba(240, 82, 82, var(--text-opacity))"
    }],
    errorText: [{
      "--text-opacity": "1",
      "color": "rgba(107, 114, 128, var(--text-opacity))"
    }],
    defaultText: [{
      "--text-opacity": "1",
      "color": "rgba(107, 114, 128, var(--text-opacity))"
    }]
  };
  return mapStyles(styles);
}

export default useMessageStyles;
