// @sajari/react-search-ui/dist/Results/components/Result/styles.esm.js
import { inferStylesObjectKeys, mapStyles } from '@sajari/react-sdk-utils';

function useResultStyles(props) {
  var appearance = props.appearance,
      isOnSale = props.isOnSale;
  var styles = inferStylesObjectKeys({
    container: [],
    imageContainer: [],
    image: [{
      "borderRadius": "0.375rem"
    }],
    content: [{
      "flex": "1 1 0%",
      "minWidth": "0"
    }],
    header: [{
      "display": "flex",
      "alignItems": "flex-start",
      "justifyContent": "space-between"
    }],
    title: [{
      "fontWeight": "500"
    }],
    subtitle: [{
      "fontSize": "0.75rem",
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))"
    }],
    description: [{
      "marginTop": "0.25rem",
      "fontSize": "0.875rem",
      "--text-opacity": "1",
      "color": "rgba(107, 114, 128, var(--text-opacity))"
    }],
    rating: [{
      "marginTop": "0.25rem",
      "> :not(template) ~ :not(template)": {
        "--space-y-reverse": 0,
        "marginTop": "calc(0.25rem * calc(1 - var(--space-y-reverse)))",
        "marginBottom": "calc(0.25rem * var(--space-y-reverse))"
      },
      "textAlign": "center"
    }],
    priceContainer: [],
    price: [isOnSale ? {
      "fontWeight": "500",
      "--text-opacity": "1",
      "color": "rgba(240, 82, 82, var(--text-opacity))"
    } : ''],
    originalPrice: [{
      "fontSize": "0.75rem",
      "--text-opacity": "1",
      "color": "rgba(159, 166, 178, var(--text-opacity))",
      "textDecoration": "line-through"
    }]
  });

  switch (appearance) {
    case 'grid':
      styles.container.push({
        "textAlign": "center"
      });
      styles.imageContainer.push({
        "display": "block",
        "marginBottom": "1rem"
      });
      styles.priceContainer.push({
        "display": "flex",
        "alignItems": "baseline",
        "justifyContent": "center",
        "> :not(template) ~ :not(template)": {
          "--space-x-reverse": 0,
          "marginRight": "calc(0.25rem * var(--space-x-reverse))",
          "marginLeft": "calc(0.25rem * calc(1 - var(--space-x-reverse)))"
        }
      });
      break;

    default:
    case 'list':
      styles.container.push({
        "display": "flex",
        "alignItems": "center",
        "width": "100%"
      });
      styles.imageContainer.push({
        "width": "6rem",
        "marginRight": "1.5rem"
      });
      styles.priceContainer.push({
        "textAlign": "right"
      });
      break;
  }

  return mapStyles(styles);
}

export default useResultStyles;
